authentication/tests.py

```diff
@@ -1,10 +1,132 @@
 from django.test import TestCase
-from authentication.models import *
-
-# Create your tests here.
-class MockTest(TestCase):
-    def testTodo(self):
-        label = "mock"
-        description = "mock"
-        obj = Todo(label=label,description=description)
-        self.assertEqual(obj.label,label)
+from authentication.models import AppUser
+from django.urls import reverse
+import json
+from django.core import mail
+from rest_framework.test import APIClient
+from .tokens import account_token
+
+REGISTER_LINK = reverse('authentication:register')
+LOGIN_LINK = reverse('authentication:login')
+EMAIL_VERIFICATION_LINK = reverse('authentication:verify_email')
+RECOVER_PASSWORD_LINK = reverse('authentication:recover_password')
+
+class RegisterTest(TestCase):
+
+    def setUp(self):
+        self.client = APIClient()
+
+    def test_register_valid(self):
+        data = {
+                "username":"user1",
+                "password":"pass1",
+                "email":"email1@email.com"
+        }
+        response = self.client.post(REGISTER_LINK, json.dumps(data), content_type='application/json')
+        self.assertEqual(response.status_code, 200)
+        self.assertTrue(AppUser.objects.filter(username='user1').exists())
+    
+    def test_email_format_not_valid(self):
+        data ={
+                "username":"user1",
+                "password":"pass1",
+                "email":"emailnotvalid"
+        }
+        response = self.client.post(REGISTER_LINK, json.dumps(data), content_type='application/json')
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.json()['msg'],"Email format is wrong.")
+        self.assertFalse(AppUser.objects.filter(username='user1').exists())
+    
+    def test_empty_input(self):
+        data = {
+                "username":"",
+                "password":"",
+                "email":""
+        }
+        response = self.client.post(REGISTER_LINK, json.dumps(data), content_type='application/json')
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.json()['msg'],"Empty input! Make sure all the fields are filled.")
+    
+    def test_username_is_already_exist(self):
+        data = {
+                "username":"user1",
+                "password":"pass1",
+                "email":"email1@email.com"
+        }
+        response = self.client.post(REGISTER_LINK, json.dumps(data), content_type='application/json')
+        self.assertEqual(response.status_code, 200)
+        response = self.client.post(REGISTER_LINK, json.dumps(data), content_type='application/json')
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.json()['msg'],"Username and/or email already taken!")
```

Comment:
Perhaps you could add if the some request body are missing, such as username, etc.

Reply:
The test_empty_input test that though. I also think that at first but it feels like duplicating the test. So i guess one test is enough

Reply:
Oh, I meant what if the request body is like this:

{
 "username":"user1",
 "password":"pass1",
}

Note that there is no email attribute there.